ocaml/typing/typetexp.ml

```diff
@@ -572,23 +572,18 @@ let get_type_param_name styp =
   | Ptyp_var name -> Some name
   | _ -> Misc.fatal_error "non-type-variable in get_type_param_name"
 
-let get_alloc_mode styp =
-  let modes, _ = Jane_syntax.Mode_expr.of_attrs styp.ptyp_attributes in
-  Typemode.transl_alloc_mode modes
-
 let rec extract_params styp =
-  let final styp =
-    [], styp, get_alloc_mode styp
-  in
   match styp.ptyp_desc with
-  | Ptyp_arrow (l, a, r) ->
-      let arg_mode = get_alloc_mode a in
+  | Ptyp_arrow (l, a, r, ma, mr) ->
+      let arg_mode = Typemode.transl_alloc_mode ma in
       let params, ret, ret_mode =
-        if Builtin_attributes.has_curry r.ptyp_attributes then final r
-        else extract_params r
+        match r.ptyp_desc with
+        | Ptyp_arrow _ when not (Builtin_attributes.has_curry r.ptyp_attributes) ->
+          extract_params r
```

Comment:
I'm not sure why we add this extra condition `Ptyp_arrow _` - this causes the `assert false` below. Can we keep the original logic?

Reply:
There's no way to extract the return mode from just the `styp` of the return type anymore (previously that info was on the `styp.ptyp_attributes`, but now it's in the `Ptype_arrow`). So this reorganization was necessary. I think the `assert false` makes sense anyways, it doesn't make sense to extract params from something that isnt a `Ptyp_arrow`

Reply:
Yeah I thought "not an arrow type" is the base case. But in fact the caller of this function already ensures `styp` is an arrow type. So you are right - this is better.

We can avoid the `assert false` and matching on `Ptyp_arrow` twice by making `extract_params` takes `l, a, r, ma, mr`. But that's probably not  worthwhile.